File: app/src/main/java/by/offvanhooijdonk/tofreedom/helper/surprise/SurpriseHelper.java
package by.offvanhooijdonk.tofreedom.helper.surprise;

import android.app.Activity;
import android.content.Context;
import android.os.Handler;
import android.support.v7.app.ActionBar;
import android.view.ViewGroup;

public class SurpriseHelper {
    private static final double FACTOR_EVENT = 0.2;
    /*private static final double FACTOR_EVENT = 1;*/

    private ViewGroup root;
    private Context ctx;
    private Activity activity;
    private ActionBar actionBar;

    public SurpriseHelper(Context context, Activity activity, ViewGroup root, ActionBar actionBar) {
        this.ctx = context;
        this.activity = activity;
        this.root = root;
        this.actionBar = actionBar;
    }

    public void startRollingEvent() {
        startRollForEvent();
    }
    

    private void pickAndRunEvent() {
        IEvent.IEventBuilder builder = EventFactory.getRandomEventBuilder();
        /*BatsEvent batsEvent = new BatsEvent.Builder()*/
        IEvent event = builder
                .actionBar(actionBar)
                .activity(activity)
                .context(ctx)
                .rootView(root)
                .build();

        event.run();
    }

    private void startRollForEvent() {
        Handler handler = new Handler();
        Runnable roll = new Runnable() {
            @Override
            public void run() {
                if (rollForEvent()) {
                    pickAndRunEvent();
                } else {
                    handler.postDelayed(this, 30 * 1000);
                }
            }
        };
        handler.post(roll);
    }

    private boolean rollForEvent() {
        return Math.random() < FACTOR_EVENT;
    }
}

File: app/src/main/java/by/offvanhooijdonk/tofreedom/helper/surprise/SurpriseHelper.kt
package by.offvanhooijdonk.tofreedom.helper.surprise

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.support.v7.app.ActionBar
import android.view.ViewGroup

class SurpriseHelper(private val ctx: Context, private val activity: Activity,
                     private val root: ViewGroup, private val actionBar: ActionBar?) {

    fun startRollingEvent() {
        startRollForEvent()
    }

    private fun pickAndRunEvent() {
        val builder = EventFactory.randomEventBuilder

        val event = builder
                .apply { actionBar?.let { actionBar(it) }}
                .activity(activity)
                .context(ctx)
                .rootView(root)
                .build()

        event.run()
    }

    private fun startRollForEvent() {
        val handler = Handler()
        val roll = object : Runnable {
            override fun run() {
                if (rollForEvent()) {
                    pickAndRunEvent()
                } else {
                    handler.postDelayed(this, 30 * 1000L)
                }
            }
        }
        handler.post(roll)
    }

    private fun rollForEvent(): Boolean {
        return Math.random() < FACTOR_EVENT
    }

    companion object {
        private const val FACTOR_EVENT = 0.2
    }
}
